sql/database.go

```diff
@@ -444,6 +683,177 @@ func (db *sqliteDatabase) Close() error {
+// are invoked upon each query. The query will fail if the interceptor returns an error.
+// The interceptor can later be removed using RemoveInterceptor with the same key.
+func (db *sqliteDatabase) Intercept(key string, fn Interceptor) {
+	db.interceptMtx.Lock()
+	defer db.interceptMtx.Unlock()
+	if db.interceptors == nil {
+		db.interceptors = make(map[string]Interceptor)
+	}
+	db.interceptors[key] = fn
+}
+
+// PopIntercept removes the interceptor function with specified key from the database.
+// If there's no such interceptor, the function does nothing.
+func (db *sqliteDatabase) RemoveInterceptor(key string) {
+	db.interceptMtx.Lock()
+	defer db.interceptMtx.Unlock()
+	delete(db.interceptors, key)
+}
+
+// vacuumInto runs VACUUM INTO on the database and saves the vacuumed
+// database at toPath.
+func (db *sqliteDatabase) vacuumInto(toPath string) error {
+	if _, err := db.Exec("VACUUM INTO ?1", func(stmt *Statement) {
+		stmt.BindText(1, toPath)
+	}, nil); err != nil {
+		return fmt.Errorf("vacuum into %s: %w", toPath, err)
+	}
+	return nil
+}
+
+// copyMigrateDB performs a copy-based migration of the database.
+// The source database is always closed by this function.
+// Upon success, the migrated database is opened.
+func (db *sqliteDatabase) copyMigrateDB(config *conf) (finalDB *sqliteDatabase, err error) {
+	defer db.Close()
+
+	dbPath, migratedPath, err := dbMigrationPaths(config.uri)
+	if err != nil {
+		return nil, err
+	}
+	if migratedPath == "" {
+		return nil, fmt.Errorf("cannot migrate database, only file DBs are supported: %s", config.uri)
+	}
+
+	// Instead of just copying the source database to the temporary migration DB, use VACUUM INTO.
+	// This is somewhat slower but achieves two goals:
+	// 1. The lock is held on the source database while it's being copied
+	// 2. If the source database has a lot of free pages for whatever reason, those
+	// are not copied, saving disk space
+	config.logger.Info("making a temporary copy of the database",
+		zap.String("path", dbPath),
+		zap.String("target", migratedPath))
+	if err := db.vacuumInto(migratedPath); err != nil {
+		return nil, errors.Join(err, deleteDB(migratedPath))
+	}
+
+	// Opening the temporary migrated DB runs the actual migrations on it.
+	// We disable vacuuming here because we're going to vacuum the temporary DB
+	// into the original one.
+	opts := []Opt{
+		WithLogger(config.logger),
+		WithConnections(1),
+		WithTemp(),
+		WithDatabaseSchema(config.schema),
+	}
+	if !config.checkSchemaDrift {
+		opts = append(opts, WithNoCheckSchemaDrift())
+	}
+	migratedDB, err := Open("file:"+migratedPath, opts...)
+	if err != nil {
+		return nil, errors.Join(
+			fmt.Errorf("process temporary DB %s: %w", migratedPath, err),
+			deleteDB(migratedPath))
+	}
+	defer migratedDB.Close()
+	tempDBReady := false
+	defer func() {
+		err = errors.Join(err, migratedDB.Close())
+		if !tempDBReady {
+			err = errors.Join(err, deleteDB(migratedPath))
+		}
+	}()
```

Comment:
I'm not sure if we even should return this errors at that point. There are only 2 options:

a) up until this point everything went well & `finalDB` is ready to be used, then failing to close or delete the temporary DB indicates to the caller that `finalDB` cannot be used
b) some other error occurred that already causes the function to return an `err`. I think in this case whatever error has already happened is more important than the two that might be returned here.

Wdyt?

Reply:
As above, leaving an incomplete database in place (e.g. due to fs becoming readonly) may be a serious nuisance b/c it may occupy a large amount of disk space, and in case if the user doesn't intend to start `go-spacemesh` again soon, it may need to be cleaned up by hand.
If the database fails to be closed for whatever reason, this may also in some cases prevent deletion on Windows, so reporting close errors here may also make sense